woocommerce/woocommerce-ios
# [Mobile Payments] Retry when a payment capture fails

Closes #4210 , closes #4228

⚠️ This PR is against the feature branch implementing support for Card Present Payments, not against develop ⚠️

Sorry, this is a long PR, but most of the diff is because I have added unit tests for the view models backing the modal alerts, which were missing. 

Also, the copy of the retry button in the error alert is truncated. We have a ticket for that: #4222 

## Changes
* Add a new view model to back an error modal alert that does not retry.
* Add logic to retry a failed payment capture: 
1. Cancelling the previous payment capture
2. Starting a new payment capture after cancelling the previous one succeeds
3. Cancelling the whole process if cancelling the previous payment fails.
* Touch the error modals to display the error localized description.

There is the potential of entering an endless loop if capturing a payment fails, and a retry fails. I think we should add a "dismiss and do not not do anything"  button to the "error with retry" modal. I logged #4227 for that.

## How to test
Testing requires:

1. Setting up a store for Card Present Payment Testing (also P91TBi-4BH-p2 for more specific instructions)
2. WCPay 2.4.0

* Checkout the branch
* Connect to a reader
* Navigate to an order eligible for CPP.
* Wait a couple of seconds to make sure there are no lingering network requests, otherwise the order details screen will transition to a "no network" error screen, which is fine, but makes testing this PR a bit annoying

* Switch on airplane mode. 
* Tap "Collect payment". An error modal should be presented, because there is no network connection. The button and the error description are 😬 . We have #4222 and #4176 for that.
* Tap "retry", still in airplane mode. There should be a new alert modal. Tapping the button in the new modal should dismiss it.
* Deactivate airplane mode. Tap Collect Payment again. The payment collection should proceed forward and succeed (and the order notes should be updated).

There is one edge case we are not covering:
* Switch on airplane mode. 
* Tap "Collect payment". An error modal should be presented, because there is no network connection. The button and the error description are 😬 . We have #4222 and #4176 for that.
* Deactivate airplane mode while the modal alert is active
* Tap retry.  The payment collection could proceed forward and succeed, but there is a second error alert. After dismissing the second alert, tapping "Collect payment" again should work. We can add a ticket for this if we want to.

Activating airplane mode was the simplest way I could think of to make the payment collection process fail. If you know of another way to make that fail, by all means, try that as well! 

Update release notes:
